// @redwoodjs/auth/dist/cjs/AuthProvider/ServerAuthProvider.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ServerAuthProvider_exports = {};
__export(ServerAuthProvider_exports, {
  ServerAuthContext: () => ServerAuthContext,
  ServerAuthProvider: () => ServerAuthProvider
});
module.exports = __toCommonJS(ServerAuthProvider_exports);
var import_react = __toESM(require("react"), 1);
var import_AuthProviderState = require("./AuthProviderState.js");
const getAuthInitialStateFromServer = () => {
  if (globalThis?.__REDWOOD__SERVER__AUTH_STATE__) {
    const initialState = {
      ...import_AuthProviderState.middlewareDefaultAuthProviderState,
      encryptedSession: null,
      ...globalThis?.__REDWOOD__SERVER__AUTH_STATE__ || {}
    };
    globalThis.__REDWOOD__SERVER__AUTH_STATE__ = null;
    return initialState;
  }
  return null;
};
const ServerAuthContext = import_react.default.createContext(
  getAuthInitialStateFromServer()
);
const ServerAuthProvider = ({
  value,
  children
}) => {
  const stringifiedAuthState = `__REDWOOD__SERVER__AUTH_STATE__ = ${JSON.stringify(
    sanitizeServerAuthState(value)
  )};`;
  return /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null, /* @__PURE__ */ import_react.default.createElement(
    "script",
    {
      id: "__REDWOOD__SERVER_AUTH_STATE__",
      dangerouslySetInnerHTML: {
        __html: stringifiedAuthState
      }
    }
  ), /* @__PURE__ */ import_react.default.createElement(ServerAuthContext.Provider, { value }, children));
};
function sanitizeServerAuthState(value) {
  const sanitizedState = { ...value };
  delete sanitizedState.cookieHeader;
  return sanitizedState;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ServerAuthContext,
  ServerAuthProvider
});
