// @redwoodjs/auth/dist/cjs/AuthProvider/useLogOut.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var useLogOut_exports = {};
__export(useLogOut_exports, {
  useLogOut: () => useLogOut
});
module.exports = __toCommonJS(useLogOut_exports);
var import_react = require("react");
const useLogOut = (authImplementation, setAuthProviderState) => {
  return (0, import_react.useCallback)(
    async (options) => {
      const logoutOutput = await authImplementation.logout(options);
      setAuthProviderState({
        userMetadata: null,
        currentUser: null,
        isAuthenticated: false,
        hasError: false,
        error: void 0,
        loading: false
      });
      return logoutOutput;
    },
    [authImplementation, setAuthProviderState]
  );
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useLogOut
});
