// @redwoodjs/cli/dist/commands/generate/dataMigration/dataMigration.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dataMigration_exports = {};
__export(dataMigration_exports, {
  aliases: () => aliases,
  builder: () => builder,
  command: () => command,
  description: () => description,
  files: () => files,
  handler: () => handler
});
module.exports = __toCommonJS(dataMigration_exports);
var import_path = __toESM(require("path"));
var import_change_case = require("change-case");
var import_fs_extra = __toESM(require("fs-extra"));
var import_listr2 = require("listr2");
var import_terminal_link = __toESM(require("terminal-link"));
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_lib = require("../../../lib");
var import_colors = __toESM(require("../../../lib/colors"));
var import_rollback = require("../../../lib/rollback");
var import_helpers = require("../helpers");
const POST_RUN_INSTRUCTIONS = `Next steps...

   ${import_colors.default.warning(
  "After writing your migration, you can run it with:"
)}

     yarn rw dataMigrate up
`;
const TEMPLATE_PATHS = {
  js: import_path.default.resolve(__dirname, "templates", "dataMigration.js.template"),
  ts: import_path.default.resolve(__dirname, "templates", "dataMigration.ts.template")
};
const files = ({ name, typescript }) => {
  const now = (/* @__PURE__ */ new Date()).toISOString();
  const timestamp = now.split(".")[0].replace(/\D/g, "");
  const basename = `${timestamp}-${(0, import_change_case.paramCase)(name)}`;
  const extension = typescript ? "ts" : "js";
  const outputFilename = basename + "." + extension;
  const outputPath = import_path.default.join((0, import_lib.getPaths)().api.dataMigrations, outputFilename);
  return {
    [outputPath]: import_fs_extra.default.readFileSync(TEMPLATE_PATHS[extension]).toString()
  };
};
const command = "data-migration <name>";
const aliases = ["dataMigration", "dm"];
const description = "Generate a data migration";
const builder = (yargs) => {
  yargs.positional("name", {
    description: "A descriptor of what this data migration does",
    type: "string"
  }).option("rollback", {
    description: "Revert all generator actions if an error occurs",
    type: "boolean",
    default: true
  }).epilogue(
    `Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#generate-datamigration"
    )}`
  );
  Object.entries(import_helpers.yargsDefaults).forEach(([option, config]) => {
    yargs.option(option, config);
  });
};
const handler = async (args) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "generate data-migration",
    force: args.force,
    rollback: args.rollback
  });
  (0, import_helpers.validateName)(args.name);
  const tasks = new import_listr2.Listr(
    [
      {
        title: "Generating data migration file...",
        task: () => {
          return (0, import_lib.writeFilesTask)(files(args));
        }
      },
      {
        title: "Next steps...",
        task: (_ctx, task) => {
          task.title = POST_RUN_INSTRUCTIONS;
        }
      }
    ].filter(Boolean),
    { rendererOptions: { collapseSubtasks: false } }
  );
  try {
    if (args.rollback && !args.force) {
      (0, import_rollback.prepareForRollback)(tasks);
    }
    await tasks.run();
  } catch (e) {
    console.log(import_colors.default.error(e.message));
    process.exit(1);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  aliases,
  builder,
  command,
  description,
  files,
  handler
});
