// @redwoodjs/api-server/dist/createServer.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var createServer_exports = {};
__export(createServer_exports, {
  createServer: () => createServer
});
module.exports = __toCommonJS(createServer_exports);
var import_fs = __toESM(require("fs"));
var import_path = __toESM(require("path"));
var import_chalk = __toESM(require("chalk"));
var import_dotenv_defaults = require("dotenv-defaults");
var import_fast_glob = __toESM(require("fast-glob"));
var import_fastify = __toESM(require("fastify"));
var import_store = require("@redwoodjs/context/dist/store");
var import_project_config = require("@redwoodjs/project-config");
var import_createServerHelpers = require("./createServerHelpers");
var import_api = require("./plugins/api");
if (!process.env.REDWOOD_ENV_FILES_LOADED) {
  (0, import_dotenv_defaults.config)({
    path: import_path.default.join((0, import_project_config.getPaths)().base, ".env"),
    defaults: import_path.default.join((0, import_project_config.getPaths)().base, ".env.defaults"),
    multiline: true
  });
  process.env.REDWOOD_ENV_FILES_LOADED = "true";
}
async function createServer(options = {}) {
  const {
    apiRootPath,
    fastifyServerOptions,
    configureApiServer,
    apiPort,
    apiHost
  } = (0, import_createServerHelpers.resolveOptions)(options);
  const serverConfigPath = import_path.default.join(
    (0, import_project_config.getPaths)().base,
    (0, import_project_config.getConfig)().api.serverConfig
  );
  if (import_fs.default.existsSync(serverConfigPath)) {
    console.warn(
      import_chalk.default.yellow(
        [
          "",
          `Ignoring \`config\` and \`configureServer\` in api/server.config.js.`,
          `Migrate them to api/src/server.{ts,js}:`,
          "",
          `\`\`\`js title="api/src/server.{ts,js}"`,
          "// Pass your config to `createServer`",
          "const server = createServer({",
          "  fastifyServerOptions: myFastifyConfig",
          "})",
          "",
          "// Then inline your `configureFastify` logic:",
          "server.register(myFastifyPlugin)",
          "```",
          ""
        ].join("\n")
      )
    );
  }
  const server = Object.assign((0, import_fastify.default)(fastifyServerOptions), {
    // `start` will get replaced further down in this file
    start: async () => {
      throw new Error("Not implemented yet");
    }
  });
  server.addHook("onRequest", (_req, _reply, done) => {
    (0, import_store.getAsyncStoreInstance)().run(/* @__PURE__ */ new Map(), done);
  });
  await server.register(import_api.redwoodFastifyAPI, {
    redwood: {
      apiRootPath,
      fastGlobOptions: {
        ignore: ["**/dist/functions/graphql.js"]
      },
      configureServer: configureApiServer
    }
  });
  const [graphqlFunctionPath] = await (0, import_fast_glob.default)("dist/functions/graphql.{ts,js}", {
    cwd: (0, import_project_config.getPaths)().api.base,
    absolute: true
  });
  if (graphqlFunctionPath) {
    const { redwoodFastifyGraphQLServer } = await import("./plugins/graphql.js");
    const { __rw_graphqlOptions } = await import(`file://${graphqlFunctionPath}`);
    await server.register(redwoodFastifyGraphQLServer, {
      redwood: {
        apiRootPath,
        graphql: __rw_graphqlOptions
      }
    });
  }
  server.addHook("onReady", (done) => {
    process.send?.("ready");
    done();
  });
  server.addHook("onListen", (done) => {
    console.log(
      `Server listening at ${import_chalk.default.magenta(
        `${server.listeningOrigin}${apiRootPath}`
      )}`
    );
    done();
  });
  server.start = (options2 = {}) => {
    return server.listen({
      ...options2,
      port: apiPort,
      host: apiHost
    });
  };
  return server;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createServer
});
