// @redwoodjs/cli/dist/commands/generate/function/function.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var function_exports = {};
__export(function_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  files: () => files,
  handler: () => handler
});
module.exports = __toCommonJS(function_exports);
var import_path = __toESM(require("path"));
var import_camelcase = __toESM(require("camelcase"));
var import_listr2 = require("listr2");
var import_terminal_link = __toESM(require("terminal-link"));
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_telemetry = require("@redwoodjs/telemetry");
var import_lib = require("../../../lib");
var import_colors = __toESM(require("../../../lib/colors"));
var import_rollback = require("../../../lib/rollback");
var import_helpers = require("../helpers");
var import_helpers2 = require("../helpers");
const files = async ({
  name,
  typescript: generateTypescript = false,
  tests: generateTests = true,
  ...rest
}) => {
  const extension = generateTypescript ? ".ts" : ".js";
  const functionName = (0, import_camelcase.default)(name);
  const outputFiles = [];
  const functionFiles = await (0, import_helpers2.templateForComponentFile)({
    name: functionName,
    componentName: functionName,
    extension,
    apiPathSection: "functions",
    generator: "function",
    templatePath: "function.ts.template",
    templateVars: { ...rest },
    outputPath: import_path.default.join(
      (0, import_lib.getPaths)().api.functions,
      functionName,
      `${functionName}${extension}`
    )
  });
  outputFiles.push(functionFiles);
  if (generateTests) {
    const testFile = await (0, import_helpers2.templateForComponentFile)({
      name: functionName,
      componentName: functionName,
      extension,
      apiPathSection: "functions",
      generator: "function",
      templatePath: "test.ts.template",
      templateVars: { ...rest },
      outputPath: import_path.default.join(
        (0, import_lib.getPaths)().api.functions,
        functionName,
        `${functionName}.test${extension}`
      )
    });
    const scenarioFile = await (0, import_helpers2.templateForComponentFile)({
      name: functionName,
      componentName: functionName,
      extension,
      apiPathSection: "functions",
      generator: "function",
      templatePath: "scenarios.ts.template",
      templateVars: { ...rest },
      outputPath: import_path.default.join(
        (0, import_lib.getPaths)().api.functions,
        functionName,
        `${functionName}.scenarios${extension}`
      )
    });
    outputFiles.push(testFile);
    outputFiles.push(scenarioFile);
  }
  return outputFiles.reduce(async (accP, [outputPath, content]) => {
    const acc = await accP;
    const template = generateTypescript ? content : await (0, import_lib.transformTSToJS)(outputPath, content);
    return {
      [outputPath]: template,
      ...acc
    };
  }, Promise.resolve({}));
};
const command = "function <name>";
const description = "Generate a Function";
const builder = (yargs) => {
  yargs.positional("name", {
    description: "Name of the Function",
    type: "string"
  }).option("rollback", {
    description: "Revert all generator actions if an error occurs",
    type: "boolean",
    default: true
  }).epilogue(
    `Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#generate-function"
    )}`
  );
  Object.entries(import_helpers.yargsDefaults).forEach(([option, config]) => {
    yargs.option(option, config);
  });
};
const handler = async ({ name, force, ...rest }) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "generate function",
    force,
    rollback: rest.rollback
  });
  (0, import_helpers2.validateName)(name);
  const tasks = new import_listr2.Listr(
    [
      {
        title: "Generating function files...",
        task: async () => {
          return (0, import_lib.writeFilesTask)(await files({ name, ...rest }), {
            overwriteExisting: force
          });
        }
      }
    ],
    { rendererOptions: { collapseSubtasks: false }, exitOnError: true }
  );
  try {
    if (rest.rollback && !force) {
      (0, import_rollback.prepareForRollback)(tasks);
    }
    await tasks.run();
    console.info("");
    console.info(import_colors.default.warning("\u26A0 Important:"));
    console.info("");
    console.info(
      import_colors.default.bold(
        "When deployed, a custom serverless function is an open API endpoint and is your responsibility to secure appropriately."
      )
    );
    console.info("");
    console.info(
      `Please consult the ${(0, import_terminal_link.default)(
        "Serverless Function Considerations",
        "https://redwoodjs.com/docs/serverless-functions#security-considerations"
      )} in the RedwoodJS documentation for more information.`
    );
    console.info("");
  } catch (e) {
    (0, import_telemetry.errorTelemetry)(process.argv, e.message);
    console.error(import_colors.default.error(e.message));
    process.exit(e?.exitCode || 1);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  files,
  handler
});
