// @redwoodjs/router/dist/rsc/RscCache.js
class RscCache {
  cache = /* @__PURE__ */ new Map();
  socket;
  sendRetries = 0;
  // Turn the cache off for now. We can turn it on later if we decide we need it
  isEnabled = false;
  constructor() {
    this.socket = new WebSocket("ws://localhost:18998");
    this.socket.addEventListener("open", () => {
    });
    this.socket.addEventListener("message", (event) => {
      if (event.data.startsWith("{")) {
        const data = JSON.parse(event.data);
        if (data.id === "rsc-cache-delete") {
          if (!this.cache.has(data.key)) {
          }
          this.cache.delete(data.key);
          this.sendUpdateToWebSocket();
        } else if (data.id === "rsc-cache-clear") {
          this.cache.clear();
          this.sendToWebSocket("update", { fullCache: {} });
        } else if (data.id === "rsc-cache-enable") {
          this.isEnabled = true;
          this.sendUpdateToWebSocket();
        } else if (data.id === "rsc-cache-disable") {
          this.isEnabled = false;
        } else if (data.id === "rsc-cache-read") {
          this.sendUpdateToWebSocket();
        }
      }
    });
  }
  get(key) {
    const value = this.cache.get(key);
    console.log("RscCache.get", key, value);
    return value;
  }
  set(key, value) {
    console.log("RscCache.set", key, value);
    if (!this.isEnabled) {
      this.cache.clear();
    }
    this.cache.set(key, value);
    value.then((_resolvedValue) => {
      this.sendToWebSocket("set", {
        updatedKey: key,
        fullCache: Object.fromEntries(
          Array.from(this.cache.entries()).map(
            // @ts-expect-error hack to get the value of a Thenable
            ([location, elementThenable]) => [location, elementThenable.value]
          )
        )
      });
    });
  }
  sendToWebSocket(action, payload) {
    if (this.socket.readyState === WebSocket?.OPEN) {
      this.sendRetries = 0;
      this.socket.send(JSON.stringify({ id: "rsc-cache-" + action, payload }));
    } else if (this.socket.readyState === WebSocket?.CONNECTING && this.sendRetries < 10) {
      const backoff = 300 + this.sendRetries * 100;
      setTimeout(() => {
        this.sendRetries++;
        this.sendToWebSocket(action, payload);
      }, backoff);
    } else if (this.sendRetries >= 10) {
      console.error("Exhausted retries to send message to WebSocket server.");
    } else {
      console.error("WebSocket connection is closed.");
    }
  }
  sendUpdateToWebSocket() {
    this.sendToWebSocket("update", {
      fullCache: Object.fromEntries(
        Array.from(this.cache.entries()).map(([location, elementThenable]) => [
          location,
          // @ts-expect-error hack to get the value of a Thenable
          elementThenable.value
        ])
      )
    });
  }
}
export {
  RscCache
};
