// @redwoodjs/cli/dist/commands/generate/script/script.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var script_exports = {};
__export(script_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  files: () => files,
  handler: () => handler
});
module.exports = __toCommonJS(script_exports);
var import_path = __toESM(require("path"));
var import_fs_extra = __toESM(require("fs-extra"));
var import_listr2 = require("listr2");
var import_terminal_link = __toESM(require("terminal-link"));
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_telemetry = require("@redwoodjs/telemetry");
var import_lib = require("../../../lib");
var import_colors = __toESM(require("../../../lib/colors"));
var import_rollback = require("../../../lib/rollback");
var import_helpers = require("../helpers");
const TEMPLATE_PATH = import_path.default.resolve(__dirname, "templates", "script.js.template");
const TSCONFIG_TEMPLATE = import_path.default.resolve(
  __dirname,
  "templates",
  "tsconfig.json.template"
);
const files = ({ name, typescript = false }) => {
  const outputFilename = `${name}.${typescript ? "ts" : "js"}`;
  const outputPath = import_path.default.join((0, import_lib.getPaths)().scripts, outputFilename);
  const scriptTsConfigPath = import_path.default.join((0, import_lib.getPaths)().scripts, "tsconfig.json");
  return {
    [outputPath]: import_fs_extra.default.readFileSync(TEMPLATE_PATH, "utf-8"),
    // Add tsconfig for type and cmd+click support if project is TS
    ...typescript && !import_fs_extra.default.existsSync(scriptTsConfigPath) && {
      [scriptTsConfigPath]: import_fs_extra.default.readFileSync(TSCONFIG_TEMPLATE, "utf-8")
    }
  };
};
const command = "script <name>";
const description = "Generate a command line script";
const builder = (yargs) => {
  yargs.positional("name", {
    description: "A descriptor of what this script does",
    type: "string"
  }).option("rollback", {
    description: "Revert all generator actions if an error occurs",
    type: "boolean",
    default: true
  }).epilogue(
    `Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#generate-script"
    )}`
  );
  Object.entries(import_helpers.yargsDefaults).forEach(([option, config]) => {
    yargs.option(option, config);
  });
};
const handler = async ({ force, ...args }) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "generate script",
    force,
    rollback: args.rollback
  });
  const POST_RUN_INSTRUCTIONS = `Next steps...

   ${import_colors.default.warning(
    "After modifying your script, you can invoke it like:"
  )}

     yarn rw exec ${args.name}

     yarn rw exec ${args.name} --param1 true
`;
  (0, import_helpers.validateName)(args.name);
  const tasks = new import_listr2.Listr(
    [
      {
        title: "Generating script file...",
        task: () => {
          return (0, import_lib.writeFilesTask)(files(args), { overwriteExisting: force });
        }
      },
      {
        title: "Next steps...",
        task: (_ctx, task) => {
          task.title = POST_RUN_INSTRUCTIONS;
        }
      }
    ].filter(Boolean),
    { rendererOptions: { collapseSubtasks: false } }
  );
  try {
    if (args.rollback && !force) {
      (0, import_rollback.prepareForRollback)(tasks);
    }
    await tasks.run();
  } catch (e) {
    (0, import_telemetry.errorTelemetry)(process.argv, e.message);
    console.log(import_colors.default.error(e.message));
    process.exit(1);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  files,
  handler
});
