// @redwoodjs/cli/dist/commands/setup/generator/generator.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var generator_exports = {};
__export(generator_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(generator_exports);
var import_path = __toESM(require("path"));
var import_fs_extra = __toESM(require("fs-extra"));
var import_terminal_link = __toESM(require("terminal-link"));
var import_cli_helpers = require("@redwoodjs/cli-helpers");
const command = "generator <name>";
const description = "Copies generator templates locally for customization";
const EXCLUDE_GENERATORS = [
  "dataMigration",
  "dbAuth",
  "generator",
  "script",
  "secret"
];
const builder = (yargs) => {
  const availableGenerators = import_fs_extra.default.readdirSync(import_path.default.join(__dirname, "../../generate"), {
    withFileTypes: true
  }).filter((dir) => dir.isDirectory() && !dir.name.match(/__/)).map((dir) => dir.name);
  yargs.positional("name", {
    description: "Name of the generator to copy templates from",
    choices: availableGenerators.filter(
      (dir) => !EXCLUDE_GENERATORS.includes(dir)
    )
  }).option("force", {
    alias: "f",
    default: false,
    description: "Overwrite existing files",
    type: "boolean"
  }).epilogue(
    `Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#setup-generator"
    )}`
  );
};
const handler = async (options) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "setup generator",
    name: options.name,
    force: options.force
  });
  const { handler: handler2 } = await import("./generatorHandler.js");
  return handler2(options);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler
});
