// @redwoodjs/cli/dist/commands/build.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var build_exports = {};
__export(build_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(build_exports);
var import_terminal_link = __toESM(require("terminal-link"));
var import_colors = __toESM(require("../lib/colors"));
var import_exit = require("../lib/exit");
var import_project = require("../lib/project");
var import_checkNodeVersion = require("../middleware/checkNodeVersion");
const command = "build [side..]";
const description = "Build for production";
const builder = (yargs) => {
  const choices = (0, import_project.sides)();
  yargs.positional("side", {
    choices,
    default: choices,
    description: "Which side(s) to build",
    type: "array"
  }).option("verbose", {
    alias: "v",
    default: false,
    description: "Print more",
    type: "boolean"
  }).option("prerender", {
    default: true,
    description: "Prerender after building web",
    type: "boolean"
  }).option("prisma", {
    type: "boolean",
    alias: "db",
    default: true,
    description: "Generate the Prisma client"
  }).middleware(() => {
    const check = (0, import_checkNodeVersion.checkNodeVersion)();
    if (check.ok) {
      return;
    }
    (0, import_exit.exitWithError)(void 0, {
      message: `${import_colors.default.error("Error")}: ${check.message}`,
      includeEpilogue: false
    });
  }).epilogue(
    `Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#build"
    )}`
  );
};
const handler = async (options) => {
  const { handler: handler2 } = await import("./buildHandler.js");
  return handler2(options);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler
});
