// @redwoodjs/cli/dist/commands/dev.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dev_exports = {};
__export(dev_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(dev_exports);
var import_terminal_link = __toESM(require("terminal-link"));
var import_colors = __toESM(require("../lib/colors"));
var import_checkNodeVersion = require("../middleware/checkNodeVersion");
const command = "dev [side..]";
const description = "Start development servers for api, and web";
const builder = (yargs) => {
  yargs.positional("side", {
    choices: ["api", "web"],
    default: ["api", "web"],
    description: "Which dev server(s) to start",
    type: "array"
  }).option("forward", {
    alias: "fwd",
    description: 'String of one or more vite dev server config options, for example: `--fwd="--port=1234 --open=false"`',
    type: "string",
    hidden: true
  }).option("generate", {
    type: "boolean",
    default: true,
    description: "Generate artifacts"
  }).option("apiDebugPort", {
    type: "number",
    description: "Port on which to expose API server debugger. If you supply the flag with no value it defaults to 18911."
  }).middleware(() => {
    const check = (0, import_checkNodeVersion.checkNodeVersion)();
    if (check.ok) {
      return;
    }
    console.warn(`${import_colors.default.warning("Warning")}: ${check.message}
`);
  }).epilogue(
    `Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#dev"
    )}`
  );
};
const handler = async (options) => {
  const { handler: handler2 } = await import("./devHandler.js");
  return handler2(options);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler
});
