// @redwoodjs/cli/dist/commands/prerender.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var prerender_exports = {};
__export(prerender_exports, {
  aliases: () => aliases,
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(prerender_exports);
const command = "prerender";
const aliases = ["render"];
const description = "Prerender pages of your Redwood app at build time";
const builder = (yargs) => {
  yargs.showHelpOnFail(false);
  yargs.option("path", {
    alias: ["p", "route"],
    description: "Router path to prerender. Especially useful for debugging",
    type: "string"
  });
  yargs.option("dry-run", {
    alias: ["d", "dryrun"],
    default: false,
    description: "Run prerender and output to console",
    type: "boolean"
  });
  yargs.option("verbose", {
    alias: "v",
    default: false,
    description: "Print more",
    type: "boolean"
  });
};
const handler = async (options) => {
  const { handler: handler2 } = await import("./prerenderHandler.js");
  return handler2(options);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  aliases,
  builder,
  command,
  description,
  handler
});
