// @redwoodjs/cli/dist/commands/test.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var test_exports = {};
__export(test_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(test_exports);
var import_terminal_link = __toESM(require("terminal-link"));
var import_colors = __toESM(require("../lib/colors"));
var import_project = require("../lib/project");
const command = "test [filter..]";
const description = "Run Jest tests. Defaults to watch mode";
const builder = (yargs) => {
  yargs.strict(false).positional("filter", {
    default: (0, import_project.sides)(),
    description: "Which side(s) to test, and/or a regular expression to match against your test files to filter by",
    type: "array"
  }).option("watch", {
    describe: "Run tests related to changed files based on hg/git. Specify the name or path to a file to focus on a specific set of tests",
    type: "boolean",
    default: true
  }).option("collect-coverage", {
    describe: "Show test coverage summary and output info to coverage directory",
    type: "boolean",
    default: false
  }).option("db-push", {
    describe: "Syncs the test database with your Prisma schema without requiring a migration. It creates a test database if it doesn't already exist.",
    type: "boolean",
    default: true
  }).epilogue(
    `For all available flags, run jest cli directly ${import_colors.default.tip(
      "yarn jest --help"
    )}

Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#test"
    )}
`
  );
};
const handler = async (options) => {
  const { handler: handler2 } = await import("./testHandler.js");
  return handler2(options);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler
});
