// @redwoodjs/cli/dist/commands/type-check.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var type_check_exports = {};
__export(type_check_exports, {
  aliases: () => aliases,
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(type_check_exports);
var import_terminal_link = __toESM(require("terminal-link"));
var import_project = require("../lib/project");
const command = "type-check [sides..]";
const aliases = ["tsc", "tc"];
const description = "Run a TypeScript compiler check on your project";
const builder = (yargs) => {
  yargs.strict(false).positional("sides", {
    default: (0, import_project.sides)(),
    description: "Which side(s) to run a typecheck on",
    type: "array"
  }).option("prisma", {
    type: "boolean",
    default: true,
    description: "Generate the Prisma client"
  }).option("generate", {
    type: "boolean",
    default: true,
    description: "Regenerate types within the project"
  }).option("verbose", {
    alias: "v",
    default: false,
    description: "Print more",
    type: "boolean"
  }).epilogue(
    `Also see the ${(0, import_terminal_link.default)(
      "Redwood CLI Reference",
      "https://redwoodjs.com/docs/cli-commands#type-check"
    )}`
  );
};
const handler = async (options) => {
  const { handler: handler2 } = await import("./type-checkHandler.js");
  return handler2(options);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  aliases,
  builder,
  command,
  description,
  handler
});
