// @redwoodjs/cli/dist/lib/runTransform.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var runTransform_exports = {};
__export(runTransform_exports, {
  runTransform: () => runTransform
});
module.exports = __toCommonJS(runTransform_exports);
var jscodeshift = __toESM(require("jscodeshift/src/Runner"));
const defaultJscodeshiftOpts = {
  // 0, 1 or 2
  verbose: 0,
  dry: false,
  // Doesn't do anything when running programmatically
  print: false,
  babel: true,
  extensions: "js,ts,jsx,tsx",
  ignorePattern: "**/node_modules/**",
  ignoreConfig: [],
  runInBand: false,
  silent: true,
  parser: "babel",
  parserConfig: {},
  // `silent` has to be `false` for this option to do anything
  failOnError: false,
  stdin: false
};
const runTransform = async ({
  transformPath,
  targetPaths,
  parser = "tsx",
  options = {}
}) => {
  if (process.env.NODE_ENV === "test" && process.env.RWJS_CWD) {
    process.chdir(process.env.RWJS_CWD);
  }
  const { output, stdoutWrite } = patchStdoutWrite();
  const result = await jscodeshift.run(transformPath, targetPaths, {
    ...defaultJscodeshiftOpts,
    parser,
    babel: process.env.NODE_ENV === "test",
    ...options
    // Putting options here lets users override all the defaults.
  });
  restoreStdoutWrite(stdoutWrite);
  let error;
  if (result.error) {
    error = output.value.split("\n").find((line) => line.startsWith("Error: "))?.slice("Error: ".length);
  }
  return {
    ...result,
    error,
    output: output.value
  };
};
function patchStdoutWrite() {
  const stdoutWrite = process.stdout.write;
  const output = {
    value: ""
  };
  process.stdout.write = (chunk) => {
    if (typeof chunk === "string") {
      output.value += chunk;
    }
    return true;
  };
  return { output, stdoutWrite };
}
function restoreStdoutWrite(stdoutWrite) {
  process.stdout.write = stdoutWrite;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  runTransform
});
