// @redwoodjs/cli/dist/commands/destroy/page/page.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var page_exports = {};
__export(page_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler,
  tasks: () => tasks
});
module.exports = __toCommonJS(page_exports);
var import_camelcase = __toESM(require("camelcase"));
var import_listr2 = require("listr2");
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_lib = require("../../../lib");
var import_colors = __toESM(require("../../../lib/colors"));
var import_helpers = require("../../generate/helpers");
var import_page = require("../../generate/page/page");
const command = "page <name> [path]";
const description = "Destroy a page and route component";
const builder = (yargs) => {
  yargs.positional("name", {
    description: "Name of the page",
    type: "string"
  });
  yargs.positional("path", {
    description: "URL path to the page. Defaults to name",
    type: "string"
  });
};
const tasks = ({ name, path }) => new import_listr2.Listr(
  [
    {
      title: "Destroying page files...",
      task: async () => {
        const p = (0, import_helpers.pathName)(path, name);
        const f = (0, import_page.files)({
          name,
          path: p,
          stories: true,
          tests: true,
          ...(0, import_page.paramVariants)(p)
        });
        return (0, import_lib.deleteFilesTask)(f);
      }
    },
    {
      title: "Cleaning up routes file...",
      task: async () => (0, import_lib.removeRoutesFromRouterTask)([(0, import_camelcase.default)(name)])
    }
  ],
  { rendererOptions: { collapseSubtasks: false }, exitOnError: true }
);
const handler = async ({ name, path }) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "destroy page"
  });
  const t = tasks({ name, path });
  try {
    await t.run();
  } catch (e) {
    console.log(import_colors.default.error(e.message));
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler,
  tasks
});
