// @redwoodjs/cli/dist/commands/setup/deploy/providers/baremetal.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var baremetal_exports = {};
__export(baremetal_exports, {
  command: () => command,
  configFilename: () => configFilename,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(baremetal_exports);
var import_path = __toESM(require("path"));
var import_listr2 = require("listr2");
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_telemetry = require("@redwoodjs/telemetry");
var import_lib = require("../../../../lib");
var import_colors = __toESM(require("../../../../lib/colors"));
var import_helpers = require("../helpers");
var import_baremetal = require("../templates/baremetal");
const command = "baremetal";
const description = "Setup Baremetal deploy";
const configFilename = "deploy.toml";
const files = [
  {
    path: import_path.default.join((0, import_lib.getPaths)().base, configFilename),
    content: import_baremetal.DEPLOY
  },
  {
    path: import_path.default.join((0, import_lib.getPaths)().base, "ecosystem.config.js"),
    content: import_baremetal.ECOSYSTEM
  },
  {
    path: import_path.default.join((0, import_lib.getPaths)().web.src, "maintenance.html"),
    content: import_baremetal.MAINTENANCE
  }
];
const notes = [
  "You are almost ready to go BAREMETAL!",
  "",
  "See https://redwoodjs.com/docs/deploy/baremetal for the remaining",
  "config and setup required before you can perform your first deploy."
];
const handler = async ({ force }) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "setup deploy baremetal",
    force
  });
  const tasks = new import_listr2.Listr(
    [
      (0, import_lib.addPackagesTask)({
        packages: ["node-ssh"],
        devDependency: true
      }),
      (0, import_helpers.addFilesTask)({
        files,
        force
      }),
      (0, import_lib.printSetupNotes)(notes)
    ],
    { rendererOptions: { collapseSubtasks: false } }
  );
  try {
    await tasks.run();
  } catch (e) {
    (0, import_telemetry.errorTelemetry)(process.argv, e.message);
    console.error(import_colors.default.error(e.message));
    process.exit(e?.exitCode || 1);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  command,
  configFilename,
  description,
  handler
});
