// @redwoodjs/cli/dist/commands/setup/deploy/providers/vercel.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vercel_exports = {};
__export(vercel_exports, {
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(vercel_exports);
var import_path = __toESM(require("path"));
var import_listr2 = require("listr2");
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_telemetry = require("@redwoodjs/telemetry");
var import_lib = require("../../../../lib");
var import_colors = __toESM(require("../../../../lib/colors"));
var import_helpers = require("../helpers");
const command = "vercel";
const description = "Setup Vercel deploy";
async function handler(options) {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "setup deploy vercel"
  });
  const tasks = new import_listr2.Listr(
    [
      (0, import_helpers.updateApiURLTask)("/api"),
      writeVercelConfigTask({ overwriteExisting: options.force }),
      (0, import_lib.printSetupNotes)(notes)
    ],
    {
      rendererOptions: { collapseSubtasks: false }
    }
  );
  try {
    await tasks.run();
  } catch (e) {
    (0, import_telemetry.errorTelemetry)(process.argv, e.message);
    console.error(import_colors.default.error(e.message));
    process.exit(e?.exitCode || 1);
  }
}
function writeVercelConfigTask({ overwriteExisting = false } = {}) {
  return {
    title: "Writing vercel.json...",
    task: (_ctx, task) => {
      (0, import_lib.writeFile)(
        import_path.default.join((0, import_lib.getPaths)().base, "vercel.json"),
        JSON.stringify(vercelConfig, null, 2),
        { overwriteExisting },
        task
      );
    }
  };
}
const vercelConfig = {
  build: {
    env: {
      ENABLE_EXPERIMENTAL_COREPACK: "1"
    }
  }
};
const notes = [
  "You are ready to deploy to Vercel!",
  "See: https://redwoodjs.com/docs/deploy#vercel-deploy"
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  command,
  description,
  handler
});
