// @redwoodjs/vite/dist/plugins/vite-plugin-rsc-reload.js
import path from "node:path";
import * as swc from "@swc/core";
function rscReloadPlugin(fn) {
  let enabled = false;
  const isClientEntry = (id, code) => {
    const ext = path.extname(id);
    if ([".ts", ".tsx", ".js", ".jsx"].includes(ext)) {
      const mod = swc.parseSync(code, {
        syntax: ext === ".ts" || ext === ".tsx" ? "typescript" : "ecmascript",
        tsx: ext === ".tsx"
      });
      for (const item of mod.body) {
        if (item.type === "ExpressionStatement" && item.expression.type === "StringLiteral" && item.expression.value === "use client") {
          return true;
        }
      }
    }
    return false;
  };
  return {
    name: "reload-plugin",
    configResolved(config) {
      if (config.mode === "development") {
        enabled = true;
      }
    },
    async handleHotUpdate(ctx) {
      if (!enabled) {
        return [];
      }
      if (ctx.modules.length && !isClientEntry(ctx.file, await ctx.read())) {
        return fn("full-reload");
      } else {
        return [];
      }
    }
  };
}
export {
  rscReloadPlugin
};
