// @redwoodjs/vite/dist/plugins/vite-plugin-rsc-routes-auto-loader.js
import path from "path";
import babelGenerator from "@babel/generator";
const generate = babelGenerator.default;
import { parse as babelParse } from "@babel/parser/index.cjs";
import babelTraverse from "@babel/traverse";
const traverse = babelTraverse.default;
import * as t from "@babel/types";
import { normalizePath } from "vite";
import {
  ensurePosixPath,
  getPaths,
  importStatementPath,
  processPagesDir
} from "@redwoodjs/project-config";
const getPathRelativeToSrc = (maybeAbsolutePath) => {
  if (!path.isAbsolute(maybeAbsolutePath)) {
    return maybeAbsolutePath;
  }
  return `./${path.relative(getPaths().web.src, maybeAbsolutePath)}`;
};
const withRelativeImports = (page) => {
  return {
    ...page,
    relativeImport: ensurePosixPath(getPathRelativeToSrc(page.importPath))
  };
};
function rscRoutesAutoLoader() {
  const routesFileId = normalizePath(getPaths().web.routes);
  const pages = processPagesDir().map(withRelativeImports);
  const duplicatePageImportNames = /* @__PURE__ */ new Set();
  const sortedPageImportNames = pages.map((page) => page.importName).sort();
  for (let i = 0; i < sortedPageImportNames.length - 1; i++) {
    if (sortedPageImportNames[i + 1] === sortedPageImportNames[i]) {
      duplicatePageImportNames.add(sortedPageImportNames[i]);
    }
  }
  if (duplicatePageImportNames.size > 0) {
    const pageNames = Array.from(duplicatePageImportNames).map((name) => `'${name}'`).join(", ");
    throw new Error(
      `Unable to find only a single file ending in 'Page.{js,jsx,ts,tsx}' in the following page directories: ${pageNames}`
    );
  }
  return {
    name: "rsc-routes-auto-loader-dev",
    transform: async function(code, id) {
      if (id !== routesFileId) {
        return null;
      }
      if (pages.length === 0) {
        return null;
      }
      const ext = path.extname(id);
      const plugins = [];
      if (ext === ".jsx") {
        plugins.push("jsx");
      }
      const ast = babelParse(code, {
        sourceType: "unambiguous",
        plugins
      });
      const importedNames = /* @__PURE__ */ new Set();
      const allImports = /* @__PURE__ */ new Map();
      const wrappers = /* @__PURE__ */ new Set();
      traverse(ast, {
        ImportDeclaration(path2) {
          const importPath = path2.node.source.value;
          if (importPath === null) {
            return;
          }
          const userImportRelativePath = getPathRelativeToSrc(
            importStatementPath(importPath)
          );
          const defaultSpecifier = path2.node.specifiers.filter(
            (specifier) => t.isImportDefaultSpecifier(specifier)
          )[0];
          if (userImportRelativePath && defaultSpecifier) {
            importedNames.add(defaultSpecifier.local.name);
          }
          path2.node.specifiers.forEach((specifier) => {
            allImports.set(specifier.local.name, path2.node);
          });
        },
        JSXElement() {
        },
        CallExpression(path2) {
          if ((t.isIdentifier(path2.node.callee, { name: "jsxs" }) || t.isIdentifier(path2.node.callee, { name: "jsx" })) && t.isIdentifier(path2.node.arguments[0]) && path2.node.arguments[0].name === "Set") {
            const jsxArgs = path2.node.arguments;
            if (t.isObjectExpression(jsxArgs[1])) {
              const wrapProp = jsxArgs[1].properties.find(
                (prop) => t.isObjectProperty(prop) && t.isIdentifier(prop.key, { name: "wrap" })
              );
              if (t.isArrayExpression(wrapProp?.value)) {
                wrapProp.value.elements.forEach((element) => {
                  if (t.isIdentifier(element)) {
                    wrappers.add(element.name);
                  }
                });
              } else if (t.isIdentifier(wrapProp?.value)) {
                wrappers.add(wrapProp.value.name);
              }
            }
          }
        }
      });
      const nonImportedPages = pages.filter(
        (page) => !importedNames.has(page.importName)
      );
      wrappers.forEach((wrapper) => {
        const wrapperImport = allImports.get(wrapper);
        if (wrapperImport) {
          wrapperImport.source.value = "@redwoodjs/router/dist/dummyComponent";
        }
      });
      for (const page of nonImportedPages) {
        ast.program.body.unshift(
          t.variableDeclaration("const", [
            t.variableDeclarator(
              t.identifier(page.constName),
              t.arrowFunctionExpression([], t.nullLiteral())
            )
          ])
        );
      }
      return generate(ast).code;
    }
  };
}
export {
  rscRoutesAutoLoader
};
