// @redwoodjs/cli/dist/commands/experimental/setupInngest.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var setupInngest_exports = {};
__export(setupInngest_exports, {
  EXPERIMENTAL_TOPIC_ID: () => EXPERIMENTAL_TOPIC_ID,
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(setupInngest_exports);
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_util = require("./util");
const command = "setup-inngest";
const description = "Setup Inngest for background, scheduled, delayed, multi-step, and fan-out jobs";
const EXPERIMENTAL_TOPIC_ID = 4866;
const builder = (yargs) => {
  yargs.option("force", {
    alias: "f",
    default: false,
    description: "Overwrite existing configuration",
    type: "boolean"
  }).epilogue((0, import_util.getEpilogue)(command, description, EXPERIMENTAL_TOPIC_ID, true));
};
const handler = async (options) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "experimental setup-inngest",
    force: options.force
  });
  const { handler: handler2 } = await import("./setupInngestHandler.js");
  return handler2(options);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EXPERIMENTAL_TOPIC_ID,
  builder,
  command,
  description,
  handler
});
