// @redwoodjs/server-store/dist/serverStore.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var serverStore_exports = {};
__export(serverStore_exports, {
  createPerRequestMap: () => createPerRequestMap,
  createServerStorage: () => createServerStorage,
  getAuthState: () => getAuthState,
  getLocation: () => getLocation,
  getRequestCookies: () => getRequestCookies,
  getRequestHeaders: () => getRequestHeaders,
  setServerAuthState: () => setServerAuthState
});
module.exports = __toCommonJS(serverStore_exports);
var import_async_hooks = require("async_hooks");
var import_cookie_jar = require("@redwoodjs/cookie-jar");
let PER_REQ_STORAGE;
const createServerStorage = () => {
  PER_REQ_STORAGE = new import_async_hooks.AsyncLocalStorage();
  return PER_REQ_STORAGE;
};
const createPerRequestMap = ({
  headers,
  fullUrl,
  serverAuthState
}) => {
  const reqStore = /* @__PURE__ */ new Map();
  const headersObj = new Headers(headers);
  reqStore.set("headers", headersObj);
  reqStore.set("fullUrl", fullUrl);
  if (serverAuthState) {
    reqStore.set("serverAuthState", serverAuthState);
  }
  return reqStore;
};
const getStore = () => {
  if (!PER_REQ_STORAGE) {
    throw new Error("Server store not initialized");
  }
  return PER_REQ_STORAGE.getStore();
};
const getRequestCookies = () => {
  const headers = getRequestHeaders();
  return new import_cookie_jar.CookieJar(headers.get("cookie"));
};
const getRequestHeaders = () => {
  return getStore()?.get("headers");
};
const getAuthState = () => {
  return getStore()?.get("serverAuthState");
};
const getLocation = () => {
  return new URL(getStore()?.get("fullUrl"));
};
const setServerAuthState = (authState) => {
  const store = getStore();
  store?.set("serverAuthState", authState);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createPerRequestMap,
  createServerStorage,
  getAuthState,
  getLocation,
  getRequestCookies,
  getRequestHeaders,
  setServerAuthState
});
