// @redwoodjs/cli/dist/commands/destroy/sdl/sdl.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var sdl_exports = {};
__export(sdl_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler,
  tasks: () => tasks
});
module.exports = __toCommonJS(sdl_exports);
var import_listr2 = require("listr2");
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_lib = require("../../../lib");
var import_colors = __toESM(require("../../../lib/colors"));
var import_schemaHelpers = require("../../../lib/schemaHelpers");
var import_sdl = require("../../generate/sdl/sdl");
const command = "sdl <model>";
const description = "Destroy a GraphQL schema and service component based on a given DB schema Model";
const builder = (yargs) => {
  yargs.positional("model", {
    description: "Model to destroy the sdl of",
    type: "string"
  });
};
const tasks = ({ model }) => new import_listr2.Listr(
  [
    {
      title: "Destroying GraphQL schema and service component files...",
      task: async () => {
        const f = await (0, import_sdl.files)({ name: model });
        return (0, import_lib.deleteFilesTask)(f);
      }
    }
  ],
  { rendererOptions: { collapseSubtasks: false }, exitOnError: true }
);
const handler = async ({ model }) => {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "destroy sdl"
  });
  try {
    const { name } = await (0, import_schemaHelpers.verifyModelName)({ name: model, isDestroyer: true });
    await tasks({ model: name }).run();
  } catch (e) {
    console.log(import_colors.default.error(e.message));
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler,
  tasks
});
