// @redwoodjs/project-config/dist/index.js
// src/config.ts
import fs2 from "fs";
import merge from "deepmerge";
import * as toml from "smol-toml";
import { env as envInterpolation } from "string-env-interpolation";

// src/findUp.ts
import fs from "fs";
import path from "path";
function findUp(file, startingDirectory = process.cwd()) {
  const possibleFilepath = path.join(startingDirectory, file);
  if (fs.existsSync(possibleFilepath)) {
    return possibleFilepath;
  }
  const parentDirectory = path.dirname(startingDirectory);
  if (parentDirectory === startingDirectory) {
    return null;
  }
  return findUp(file, parentDirectory);
}

// src/configPath.ts
var CONFIG_FILE_NAME = "redwood.toml";
var getConfigPathCache = /* @__PURE__ */ new Map();
var getConfigPath = (cwd = process.env.RWJS_CWD ?? process.cwd()) => {
  if (getConfigPathCache.has(cwd)) {
    return getConfigPathCache.get(cwd);
  }
  const configPath = findUp(CONFIG_FILE_NAME, cwd);
  if (!configPath) {
    throw new Error(
      `Could not find a "${CONFIG_FILE_NAME}" file, are you sure you're in a Redwood project?`
    );
  }
  getConfigPathCache.set(cwd, configPath);
  return configPath;
};

// src/config.ts
var TargetEnum = /* @__PURE__ */ ((TargetEnum2) => {
  TargetEnum2["NODE"] = "node";
  TargetEnum2["BROWSER"] = "browser";
  TargetEnum2["REACT_NATIVE"] = "react-native";
  TargetEnum2["ELECTRON"] = "electron";
  return TargetEnum2;
})(TargetEnum || {});
var DEFAULT_CONFIG = {
  web: {
    title: "Redwood App",
    port: 8910,
    path: "./web",
    target: "browser" /* BROWSER */,
    includeEnvironmentVariables: [],
    apiUrl: "/.redwood/functions",
    fastRefresh: true,
    a11y: true,
    sourceMap: false
  },
  api: {
    title: "Redwood App",
    port: 8911,
    path: "./api",
    target: "node" /* NODE */,
    schemaPath: "./api/db/schema.prisma",
    serverConfig: "./api/server.config.js",
    debugPort: 18911
  },
  graphql: {
    fragments: false,
    trustedDocuments: false,
    includeScalars: { File: true }
  },
  browser: {
    open: false
  },
  generate: {
    tests: true,
    stories: true,
    nestScaffoldByModel: true
  },
  notifications: {
    versionUpdates: []
  },
  studio: {
    basePort: 4318,
    graphiql: {
      authImpersonation: {
        authProvider: void 0,
        userId: void 0,
        email: void 0,
        jwtSecret: "secret"
      }
    }
  },
  experimental: {
    opentelemetry: {
      enabled: false,
      wrapApi: true
    },
    cli: {
      autoInstall: true,
      plugins: [
        {
          package: "@redwoodjs/cli-storybook-vite"
        },
        {
          package: "@redwoodjs/cli-data-migrate"
        }
      ]
    },
    useSDLCodeGenForGraphQLTypes: false,
    streamingSsr: {
      enabled: false
    },
    rsc: {
      enabled: false
    },
    realtime: {
      enabled: false
    },
    reactCompiler: {
      enabled: false,
      lintOnly: false
    }
  }
};
var getConfig = (configPath = getConfigPath()) => {
  try {
    return merge(DEFAULT_CONFIG, getRawConfig(configPath));
  } catch (e) {
    throw new Error(`Could not parse "${configPath}": ${e}`);
  }
};
function getRawConfig(configPath = getConfigPath()) {
  try {
    return toml.parse(envInterpolation(fs2.readFileSync(configPath, "utf8")));
  } catch (e) {
    throw new Error(`Could not parse "${configPath}": ${e}`);
  }
}

// src/paths.ts
import fs3 from "fs";
import path2 from "path";
import fg from "fast-glob";
var PATH_API_DIR_FUNCTIONS = "api/src/functions";
var PATH_RW_SCRIPTS = "scripts";
var PATH_API_DIR_GRAPHQL = "api/src/graphql";
var PATH_API_DIR_CONFIG = "api/src/config";
var PATH_API_DIR_MODELS = "api/src/models";
var PATH_API_DIR_JOBS = "api/src/jobs";
var PATH_API_DIR_LIB = "api/src/lib";
var PATH_API_DIR_GENERATORS = "api/generators";
var PATH_API_DIR_SERVICES = "api/src/services";
var PATH_API_DIR_DIRECTIVES = "api/src/directives";
var PATH_API_DIR_SUBSCRIPTIONS = "api/src/subscriptions";
var PATH_API_DIR_SRC = "api/src";
var PATH_API_DIR_DIST = "api/dist";
var PATH_WEB_ROUTES = "web/src/Routes";
var PATH_WEB_DIR_LAYOUTS = "web/src/layouts/";
var PATH_WEB_DIR_PAGES = "web/src/pages/";
var PATH_WEB_DIR_COMPONENTS = "web/src/components";
var PATH_WEB_DIR_STORYBOOK_CONFIG = "web/.storybook";
var PATH_WEB_DIR_SRC = "web/src";
var PATH_WEB_DIR_SRC_APP = "web/src/App";
var PATH_WEB_DIR_SRC_DOCUMENT = "web/src/Document";
var PATH_WEB_INDEX_HTML = "web/src/index.html";
var PATH_WEB_DIR_GENERATORS = "web/generators";
var PATH_WEB_DIR_CONFIG = "web/config";
var PATH_WEB_DIR_CONFIG_VITE = "web/vite.config";
var PATH_WEB_DIR_ENTRY_CLIENT = "web/src/entry.client";
var PATH_WEB_DIR_ENTRY_SERVER = "web/src/entry.server";
var PATH_WEB_DIR_GRAPHQL = "web/src/graphql";
var PATH_WEB_DIR_CONFIG_POSTCSS = "web/config/postcss.config.js";
var PATH_WEB_DIR_CONFIG_STORYBOOK_CONFIG = "web/config/storybook.config.js";
var PATH_WEB_DIR_CONFIG_STORYBOOK_PREVIEW = "web/config/storybook.preview";
var PATH_WEB_DIR_CONFIG_STORYBOOK_MANAGER = "web/config/storybook.manager.js";
var PATH_WEB_DIR_DIST = "web/dist";
var PATH_WEB_DIR_DIST_BROWSER = "web/dist/browser";
var PATH_WEB_DIR_DIST_RSC = "web/dist/rsc";
var PATH_WEB_DIR_DIST_SSR = "web/dist/ssr";
var PATH_WEB_DIR_DIST_SSR_ENTRY_SERVER = "web/dist/ssr/entry.server.mjs";
var PATH_WEB_DIR_DIST_SSR_DOCUMENT = "web/dist/ssr/Document.mjs";
var PATH_WEB_DIR_DIST_SSR_ROUTEHOOKS = "web/dist/ssr/routeHooks";
var PATH_WEB_DIR_DIST_RSC_ENTRIES = "web/dist/rsc/entries.mjs";
var PATH_WEB_DIR_ROUTE_MANIFEST = "web/dist/ssr/route-manifest.json";
var getBaseDir = (configPath = getConfigPath()) => {
  return path2.dirname(configPath);
};
var getBaseDirFromFile = (file) => {
  return getBaseDir(getConfigPath(path2.dirname(file)));
};
var resolveFile = (filePath, extensions = [".js", ".tsx", ".ts", ".jsx", ".mjs", ".mts"]) => {
  for (const extension of extensions) {
    const p = `${filePath}${extension}`;
    if (fs3.existsSync(p)) {
      return p;
    }
  }
  return null;
};
var getPathsCache = /* @__PURE__ */ new Map();
var getPaths = (BASE_DIR = getBaseDir()) => {
  if (getPathsCache.has(BASE_DIR)) {
    return getPathsCache.get(BASE_DIR);
  }
  const routes = resolveFile(path2.join(BASE_DIR, PATH_WEB_ROUTES));
  const { schemaPath } = getConfig(getConfigPath(BASE_DIR)).api;
  const schemaDir = path2.dirname(schemaPath);
  const viteConfig = resolveFile(
    path2.join(BASE_DIR, PATH_WEB_DIR_CONFIG_VITE)
  );
  const paths = {
    base: BASE_DIR,
    generated: {
      base: path2.join(BASE_DIR, ".redwood"),
      schema: path2.join(BASE_DIR, ".redwood/schema.graphql"),
      types: {
        includes: path2.join(BASE_DIR, ".redwood/types/includes"),
        mirror: path2.join(BASE_DIR, ".redwood/types/mirror")
      },
      prebuild: path2.join(BASE_DIR, ".redwood/prebuild")
    },
    scripts: path2.join(BASE_DIR, PATH_RW_SCRIPTS),
    api: {
      base: path2.join(BASE_DIR, "api"),
      dataMigrations: path2.join(BASE_DIR, schemaDir, "dataMigrations"),
      db: path2.join(BASE_DIR, schemaDir),
      dbSchema: path2.join(BASE_DIR, schemaPath),
      functions: path2.join(BASE_DIR, PATH_API_DIR_FUNCTIONS),
      graphql: path2.join(BASE_DIR, PATH_API_DIR_GRAPHQL),
      lib: path2.join(BASE_DIR, PATH_API_DIR_LIB),
      generators: path2.join(BASE_DIR, PATH_API_DIR_GENERATORS),
      config: path2.join(BASE_DIR, PATH_API_DIR_CONFIG),
      services: path2.join(BASE_DIR, PATH_API_DIR_SERVICES),
      directives: path2.join(BASE_DIR, PATH_API_DIR_DIRECTIVES),
      subscriptions: path2.join(BASE_DIR, PATH_API_DIR_SUBSCRIPTIONS),
      src: path2.join(BASE_DIR, PATH_API_DIR_SRC),
      dist: path2.join(BASE_DIR, PATH_API_DIR_DIST),
      types: path2.join(BASE_DIR, "api/types"),
      models: path2.join(BASE_DIR, PATH_API_DIR_MODELS),
      mail: path2.join(BASE_DIR, PATH_API_DIR_SRC, "mail"),
      jobs: path2.join(path2.join(BASE_DIR, PATH_API_DIR_JOBS)),
      distJobs: path2.join(path2.join(BASE_DIR, PATH_API_DIR_DIST, "jobs")),
      jobsConfig: resolveFile(path2.join(BASE_DIR, PATH_API_DIR_LIB, "jobs")),
      distJobsConfig: resolveFile(
        path2.join(BASE_DIR, PATH_API_DIR_DIST, "lib", "jobs")
      ),
      logger: resolveFile(path2.join(BASE_DIR, PATH_API_DIR_LIB, "logger"))
    },
    web: {
      routes,
      base: path2.join(BASE_DIR, "web"),
      pages: path2.join(BASE_DIR, PATH_WEB_DIR_PAGES),
      components: path2.join(BASE_DIR, PATH_WEB_DIR_COMPONENTS),
      layouts: path2.join(BASE_DIR, PATH_WEB_DIR_LAYOUTS),
      src: path2.join(BASE_DIR, PATH_WEB_DIR_SRC),
      storybook: path2.join(BASE_DIR, PATH_WEB_DIR_STORYBOOK_CONFIG),
      generators: path2.join(BASE_DIR, PATH_WEB_DIR_GENERATORS),
      app: resolveFile(path2.join(BASE_DIR, PATH_WEB_DIR_SRC_APP)),
      document: resolveFile(
        path2.join(BASE_DIR, PATH_WEB_DIR_SRC_DOCUMENT)
      ),
      html: path2.join(BASE_DIR, PATH_WEB_INDEX_HTML),
      config: path2.join(BASE_DIR, PATH_WEB_DIR_CONFIG),
      viteConfig,
      postcss: path2.join(BASE_DIR, PATH_WEB_DIR_CONFIG_POSTCSS),
      storybookConfig: path2.join(
        BASE_DIR,
        PATH_WEB_DIR_CONFIG_STORYBOOK_CONFIG
      ),
      storybookPreviewConfig: resolveFile(
        path2.join(BASE_DIR, PATH_WEB_DIR_CONFIG_STORYBOOK_PREVIEW)
      ),
      storybookManagerConfig: path2.join(
        BASE_DIR,
        PATH_WEB_DIR_CONFIG_STORYBOOK_MANAGER
      ),
      dist: path2.join(BASE_DIR, PATH_WEB_DIR_DIST),
      distBrowser: path2.join(BASE_DIR, PATH_WEB_DIR_DIST_BROWSER),
      distRsc: path2.join(BASE_DIR, PATH_WEB_DIR_DIST_RSC),
      distSsr: path2.join(BASE_DIR, PATH_WEB_DIR_DIST_SSR),
      distSsrDocument: path2.join(BASE_DIR, PATH_WEB_DIR_DIST_SSR_DOCUMENT),
      distSsrEntryServer: path2.join(
        BASE_DIR,
        PATH_WEB_DIR_DIST_SSR_ENTRY_SERVER
      ),
      distRouteHooks: path2.join(BASE_DIR, PATH_WEB_DIR_DIST_SSR_ROUTEHOOKS),
      distRscEntries: path2.join(BASE_DIR, PATH_WEB_DIR_DIST_RSC_ENTRIES),
      routeManifest: path2.join(BASE_DIR, PATH_WEB_DIR_ROUTE_MANIFEST),
      types: path2.join(BASE_DIR, "web/types"),
      entryClient: resolveFile(path2.join(BASE_DIR, PATH_WEB_DIR_ENTRY_CLIENT)),
      // new vite/stream entry point for client
      entryServer: resolveFile(path2.join(BASE_DIR, PATH_WEB_DIR_ENTRY_SERVER)),
      graphql: path2.join(BASE_DIR, PATH_WEB_DIR_GRAPHQL)
    }
  };
  fs3.mkdirSync(paths.generated.types.includes, { recursive: true });
  fs3.mkdirSync(paths.generated.types.mirror, { recursive: true });
  getPathsCache.set(BASE_DIR, paths);
  return paths;
};
var getRouteHookForPage = (pagePath) => {
  if (!pagePath) {
    return null;
  }
  return fg.sync("*.routeHooks.{js,ts,tsx,jsx}", {
    absolute: true,
    cwd: path2.dirname(pagePath)
    // the page's folder
  }).at(0) || null;
};
var getAppRouteHook = (forProd = false) => {
  const rwPaths = getPaths();
  if (forProd) {
    const distAppRouteHook = path2.join(
      rwPaths.web.distRouteHooks,
      "App.routeHooks.js"
    );
    try {
      fs3.statSync(distAppRouteHook).isFile();
      return distAppRouteHook;
    } catch {
      return null;
    }
  }
  return resolveFile(path2.join(rwPaths.web.src, "App.routeHooks"));
};
var processPagesDir = (webPagesDir = getPaths().web.pages) => {
  const pagePaths = fg.sync("**/*Page.{js,jsx,ts,tsx}", {
    cwd: webPagesDir,
    ignore: ["node_modules"]
  });
  return pagePaths.map((pagePath) => {
    const p = path2.parse(pagePath);
    const importName = p.dir.replace(/\//g, "");
    const importPath = importStatementPath(
      path2.join(webPagesDir, p.dir, p.name)
    );
    const importStatement = `const ${importName} = { name: '${importName}', loader: import('${importPath}') }`;
    return {
      importName,
      constName: importName,
      importPath,
      path: path2.join(webPagesDir, pagePath),
      importStatement
    };
  });
};
var ensurePosixPath = (path3) => {
  let posixPath = path3;
  if (process.platform === "win32") {
    if (/^[A-Z]:\\/.test(path3)) {
      const drive = path3[0].toLowerCase();
      posixPath = `/${drive}/${path3.substring(3)}`;
    }
    posixPath = posixPath.replace(/\\/g, "/");
  }
  return posixPath;
};
var importStatementPath = (path3) => {
  let importPath = path3;
  if (process.platform === "win32") {
    importPath = importPath.replaceAll("\\", "/");
  }
  return importPath;
};
function packageJsonIsEsm(packageJsonPath) {
  const packageJsonContents = JSON.parse(
    fs3.readFileSync(packageJsonPath, "utf-8")
  );
  return packageJsonContents.type === "module";
}
function projectRootIsEsm() {
  return packageJsonIsEsm(path2.join(getPaths().base, "package.json"));
}
function projectSideIsEsm(side) {
  const redwoodProjectPaths = getPaths();
  return packageJsonIsEsm(
    path2.join(redwoodProjectPaths[side].base, "package.json")
  );
}
function projectIsEsm() {
  if (!projectRootIsEsm()) {
    return false;
  }
  for (const side of ["api", "web"]) {
    if (!projectSideIsEsm(side)) {
      return false;
    }
  }
  return true;
}
var isTypeScriptProject = () => {
  const paths = getPaths();
  return fs3.existsSync(path2.join(paths.web.base, "tsconfig.json")) || fs3.existsSync(path2.join(paths.api.base, "tsconfig.json"));
};
export {
  TargetEnum,
  ensurePosixPath,
  findUp,
  getAppRouteHook,
  getBaseDir,
  getBaseDirFromFile,
  getConfig,
  getConfigPath,
  getPaths,
  getRawConfig,
  getRouteHookForPage,
  importStatementPath,
  isTypeScriptProject,
  processPagesDir,
  projectIsEsm,
  projectRootIsEsm,
  projectSideIsEsm,
  resolveFile
};
