// @redwoodjs/cli/dist/commands/generate/component/component.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var component_exports = {};
__export(component_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  files: () => files,
  handler: () => handler
});
module.exports = __toCommonJS(component_exports);
var import_lib = require("../../../lib");
var import_helpers = require("../helpers");
const REDWOOD_WEB_PATH_NAME = "components";
const files = async ({ name, typescript = false, ...options }) => {
  const extension = typescript ? ".tsx" : ".jsx";
  const componentFile = await (0, import_helpers.templateForComponentFile)({
    name,
    webPathSection: REDWOOD_WEB_PATH_NAME,
    extension,
    generator: "component",
    templatePath: "component.tsx.template"
  });
  const testFile = await (0, import_helpers.templateForComponentFile)({
    name,
    extension: `.test${extension}`,
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: "component",
    templatePath: "test.tsx.template"
  });
  const storiesFile = await (0, import_helpers.templateForComponentFile)({
    name,
    extension: `.stories${extension}`,
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: "component",
    // Using two different template files here because we have a TS-specific
    // information in a comment in the .tsx template
    templatePath: typescript ? "stories.tsx.template" : "stories.jsx.template"
  });
  const files2 = [componentFile];
  if (options.stories) {
    files2.push(storiesFile);
  }
  if (options.tests) {
    files2.push(testFile);
  }
  return files2.reduce(async (accP, [outputPath, content]) => {
    const acc = await accP;
    const template = typescript ? content : await (0, import_lib.transformTSToJS)(outputPath, content);
    return {
      [outputPath]: template,
      ...acc
    };
  }, Promise.resolve({}));
};
const description = "Generate a component";
const { command, builder, handler } = (0, import_helpers.createYargsForComponentGeneration)({
  componentName: "component",
  filesFn: files
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  files,
  handler
});
