// @redwoodjs/graphql-server/dist/plugins/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var plugins_exports = {};
__export(plugins_exports, {
  useArmor: () => import_useArmor.useArmor,
  useRedwoodAuthContext: () => import_useRedwoodAuthContext.useRedwoodAuthContext,
  useRedwoodDirective: () => import_useRedwoodDirective.useRedwoodDirective,
  useRedwoodError: () => import_useRedwoodError.useRedwoodError,
  useRedwoodGlobalContextSetter: () => import_useRedwoodGlobalContextSetter.useRedwoodGlobalContextSetter,
  useRedwoodLogger: () => import_useRedwoodLogger.useRedwoodLogger,
  useRedwoodOpenTelemetry: () => import_useRedwoodOpenTelemetry.useRedwoodOpenTelemetry,
  useRedwoodPopulateContext: () => import_useRedwoodPopulateContext.useRedwoodPopulateContext,
  useRedwoodTrustedDocuments: () => import_useRedwoodTrustedDocuments.useRedwoodTrustedDocuments
});
module.exports = __toCommonJS(plugins_exports);
var import_useArmor = require("./useArmor");
var import_useRedwoodAuthContext = require("./useRedwoodAuthContext");
var import_useRedwoodDirective = require("./useRedwoodDirective");
var import_useRedwoodError = require("./useRedwoodError");
var import_useRedwoodGlobalContextSetter = require("./useRedwoodGlobalContextSetter");
var import_useRedwoodLogger = require("./useRedwoodLogger");
var import_useRedwoodTrustedDocuments = require("./useRedwoodTrustedDocuments");
var import_useRedwoodPopulateContext = require("./useRedwoodPopulateContext");
var import_useRedwoodOpenTelemetry = require("./useRedwoodOpenTelemetry");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useArmor,
  useRedwoodAuthContext,
  useRedwoodDirective,
  useRedwoodError,
  useRedwoodGlobalContextSetter,
  useRedwoodLogger,
  useRedwoodOpenTelemetry,
  useRedwoodPopulateContext,
  useRedwoodTrustedDocuments
});
