// @redwoodjs/cli/dist/commands/setup/ui/libraries/chakra-ui.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chakra_ui_exports = {};
__export(chakra_ui_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(chakra_ui_exports);
var import_path = __toESM(require("path"));
var import_execa = __toESM(require("execa"));
var import_listr2 = require("listr2");
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_lib = require("../../../../lib");
var import_colors = __toESM(require("../../../../lib/colors"));
var import_configureStorybook = __toESM(require("../../../../lib/configureStorybook.js"));
var import_extendFile = require("../../../../lib/extendFile");
const command = "chakra-ui";
const description = "Set up Chakra UI";
function builder(yargs) {
  yargs.option("force", {
    alias: "f",
    default: false,
    description: "Overwrite existing configuration",
    type: "boolean"
  });
  yargs.option("install", {
    alias: "i",
    default: true,
    description: "Install packages",
    type: "boolean"
  });
}
const CHAKRA_THEME_AND_COMMENTS = `// This object will be used to override Chakra-UI theme defaults.
// See https://chakra-ui.com/docs/styled-system/theming/theme for theming options
const theme = {}
export default theme
`;
async function handler({ force, install }) {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "setup ui chakra-ui",
    force,
    install
  });
  const rwPaths = (0, import_lib.getPaths)();
  const packages = [
    "@chakra-ui/react@^2",
    "@emotion/react@^11",
    "@emotion/styled@^11",
    "framer-motion@^9"
  ];
  const tasks = new import_listr2.Listr(
    [
      {
        title: "Installing packages...",
        skip: () => !install,
        task: () => {
          return new import_listr2.Listr(
            [
              {
                title: `Install ${packages.join(", ")}`,
                task: async () => {
                  await (0, import_execa.default)("yarn", ["workspace", "web", "add", ...packages]);
                }
              }
            ],
            { rendererOptions: { collapseSubtasks: false } }
          );
        }
      },
      {
        title: "Setting up Chakra UI...",
        skip: () => (0, import_extendFile.fileIncludes)(rwPaths.web.app, "ChakraProvider"),
        task: () => (0, import_extendFile.extendJSXFile)(rwPaths.web.app, {
          insertComponent: {
            name: "ChakraProvider",
            props: { theme: "extendedTheme" },
            within: "RedwoodProvider",
            insertBefore: "<ColorModeScript />"
          },
          imports: [
            "import { ChakraProvider, ColorModeScript, extendTheme } from '@chakra-ui/react'",
            "import * as theme from 'config/chakra.config'"
          ],
          moduleScopeLines: ["const extendedTheme = extendTheme(theme)"]
        })
      },
      {
        title: `Creating Theme File...`,
        task: () => {
          (0, import_lib.writeFile)(
            import_path.default.join(rwPaths.web.config, "chakra.config.js"),
            CHAKRA_THEME_AND_COMMENTS,
            { overwriteExisting: force }
          );
        }
      },
      {
        title: "Configure Storybook...",
        // skip this task if the user's storybook config already includes "withChakra"
        skip: () => (0, import_extendFile.fileIncludes)(rwPaths.web.storybookConfig, "withChakra"),
        task: async () => (0, import_configureStorybook.default)(
          import_path.default.join(
            __dirname,
            "..",
            "templates",
            "chakra.storybook.preview.tsx.template"
          )
        )
      }
    ],
    { rendererOptions: { collapseSubtasks: false } }
  );
  try {
    await tasks.run();
  } catch (e) {
    console.error(import_colors.default.error(e.message));
    process.exit(e?.exitCode || 1);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler
});
