// @redwoodjs/cli/dist/commands/setup/server-file/serverFileHandler.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var serverFileHandler_exports = {};
__export(serverFileHandler_exports, {
  handler: () => handler,
  setupServerFileTasks: () => setupServerFileTasks
});
module.exports = __toCommonJS(serverFileHandler_exports);
var import_path = __toESM(require("path"));
var import_fs_extra = __toESM(require("fs-extra"));
var import_listr2 = require("listr2");
var import_cli_helpers = require("@redwoodjs/cli-helpers");
var import_telemetry = require("@redwoodjs/telemetry");
var import_lib = require("../../../lib");
var import_colors = __toESM(require("../../../lib/colors"));
var import_project = require("../../../lib/project");
const { version } = JSON.parse(
  import_fs_extra.default.readFileSync(import_path.default.resolve(__dirname, "../../../../package.json"), "utf-8")
);
function setupServerFileTasks({ force = false } = {}) {
  return [
    {
      title: "Adding the server file...",
      task: async () => {
        const ts = (0, import_project.isTypeScriptProject)();
        const serverFilePath = import_path.default.join(
          (0, import_lib.getPaths)().api.src,
          `server.${ts ? "ts" : "js"}`
        );
        const serverFileTemplateContent = import_fs_extra.default.readFileSync(
          import_path.default.join(__dirname, "templates", "server.ts.template"),
          "utf-8"
        );
        const setupScriptContent = ts ? serverFileTemplateContent : await (0, import_lib.transformTSToJS)(serverFilePath, serverFileTemplateContent);
        return [
          (0, import_lib.writeFile)(serverFilePath, setupScriptContent, {
            overwriteExisting: force
          })
        ];
      }
    },
    (0, import_cli_helpers.addApiPackages)([`@redwoodjs/api-server@${version}`])
  ];
}
async function handler({ force, verbose }) {
  const tasks = new import_listr2.Listr(setupServerFileTasks({ force }), {
    rendererOptions: { collapseSubtasks: false, persistentOutput: true },
    renderer: verbose ? "verbose" : "default"
  });
  try {
    await tasks.run();
  } catch (e) {
    (0, import_telemetry.errorTelemetry)(process.argv, e.message);
    console.error(import_colors.default.error(e.message));
    process.exit(e?.exitCode || 1);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  handler,
  setupServerFileTasks
});
