// @redwoodjs/cli/dist/commands/generate/realtime/realtime.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var realtime_exports = {};
__export(realtime_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler
});
module.exports = __toCommonJS(realtime_exports);
var import_cli_helpers = require("@redwoodjs/cli-helpers");
const command = "realtime <name>";
const description = "Generate a subscription or live query used with RedwoodJS Realtime";
function builder(yargs) {
  yargs.positional("name", {
    type: "string",
    description: "Name of the realtime event to setup. This should be a type or model name like: Widget, Sprocket, etc.",
    demandOption: true
  }).option("type", {
    alias: "t",
    type: "string",
    choices: ["liveQuery", "subscription"],
    description: "Type of realtime event to setup"
  }).option("force", {
    alias: "f",
    default: false,
    description: "Overwrite existing configuration",
    type: "boolean"
  }).option("verbose", {
    alias: "v",
    default: false,
    description: "Print more logs",
    type: "boolean"
  });
}
async function handler(options) {
  (0, import_cli_helpers.recordTelemetryAttributes)({
    command: "generate realtime",
    type: options.type,
    force: options.force,
    verbose: options.verbose
  });
  const { handler: handler2 } = await import("./realtimeHandler.js");
  return handler2(options);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler
});
