// @redwoodjs/babel-config/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  BABEL_PLUGIN_TRANSFORM_RUNTIME_OPTIONS: () => import_api.BABEL_PLUGIN_TRANSFORM_RUNTIME_OPTIONS,
  CORE_JS_VERSION: () => import_common.CORE_JS_VERSION,
  RUNTIME_CORE_JS_VERSION: () => import_common.RUNTIME_CORE_JS_VERSION,
  TARGETS_NODE: () => import_api.TARGETS_NODE,
  getApiSideBabelConfigPath: () => import_api.getApiSideBabelConfigPath,
  getApiSideBabelPlugins: () => import_api.getApiSideBabelPlugins,
  getApiSideBabelPresets: () => import_api.getApiSideBabelPresets,
  getApiSideDefaultBabelConfig: () => import_api.getApiSideDefaultBabelConfig,
  getCommonPlugins: () => import_common.getCommonPlugins,
  getPathsFromConfig: () => import_common.getPathsFromTypeScriptConfig,
  getRouteHookBabelPlugins: () => import_common.getRouteHookBabelPlugins,
  getWebSideBabelConfigPath: () => import_web.getWebSideBabelConfigPath,
  getWebSideBabelPlugins: () => import_web.getWebSideBabelPlugins,
  getWebSideBabelPresets: () => import_web.getWebSideBabelPresets,
  getWebSideDefaultBabelConfig: () => import_web.getWebSideDefaultBabelConfig,
  getWebSideOverrides: () => import_web.getWebSideOverrides,
  parseTypeScriptConfigFiles: () => import_common.parseTypeScriptConfigFiles,
  registerApiSideBabelHook: () => import_api.registerApiSideBabelHook,
  registerBabel: () => import_common.registerBabel,
  registerWebSideBabelHook: () => import_web.registerWebSideBabelHook,
  transformWithBabel: () => import_api.transformWithBabel
});
module.exports = __toCommonJS(src_exports);
var import_api = require("./api");
var import_web = require("./web");
var import_common = require("./common");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BABEL_PLUGIN_TRANSFORM_RUNTIME_OPTIONS,
  CORE_JS_VERSION,
  RUNTIME_CORE_JS_VERSION,
  TARGETS_NODE,
  getApiSideBabelConfigPath,
  getApiSideBabelPlugins,
  getApiSideBabelPresets,
  getApiSideDefaultBabelConfig,
  getCommonPlugins,
  getPathsFromConfig,
  getRouteHookBabelPlugins,
  getWebSideBabelConfigPath,
  getWebSideBabelPlugins,
  getWebSideBabelPresets,
  getWebSideDefaultBabelConfig,
  getWebSideOverrides,
  parseTypeScriptConfigFiles,
  registerApiSideBabelHook,
  registerBabel,
  registerWebSideBabelHook,
  transformWithBabel
});
