// @redwoodjs/cli/dist/commands/setup/deploy/templates/flightcontrol.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var flightcontrol_exports = {};
__export(flightcontrol_exports, {
  databaseEnvVariables: () => databaseEnvVariables,
  flightcontrolConfig: () => flightcontrolConfig,
  mysqlDatabaseService: () => mysqlDatabaseService,
  postgresDatabaseService: () => postgresDatabaseService
});
module.exports = __toCommonJS(flightcontrol_exports);
const flightcontrolConfig = {
  $schema: "https://app.flightcontrol.dev/schema.json",
  environments: [
    {
      id: "development",
      name: "Development",
      region: "us-east-1",
      source: {
        branch: "main"
      },
      services: [
        {
          id: "redwood-api",
          name: "Redwood API",
          type: "web",
          buildType: "nixpacks",
          cpu: 0.5,
          memory: 1,
          buildCommand: "yarn rw deploy flightcontrol api",
          startCommand: "yarn rw deploy flightcontrol api --serve",
          port: 8911,
          healthCheckPath: "/graphql/health",
          ci: {
            type: "ec2"
          },
          envVariables: {
            REDWOOD_WEB_URL: {
              fromService: { id: "redwood-web", value: "origin" }
            }
          }
        },
        {
          id: "redwood-web",
          name: "Redwood Web",
          type: "static",
          buildType: "nixpacks",
          singlePageApp: true,
          buildCommand: "yarn rw deploy flightcontrol web",
          outputDirectory: "web/dist",
          ci: {
            type: "ec2"
          },
          envVariables: {
            REDWOOD_API_URL: {
              fromService: { id: "redwood-api", value: "origin" }
            }
          }
        }
      ]
    }
  ]
};
const postgresDatabaseService = {
  id: "db",
  name: "Database",
  type: "rds",
  engine: "postgres",
  engineVersion: "16",
  instanceSize: "db.t4g.micro",
  port: 5432,
  storage: 20,
  private: false
};
const mysqlDatabaseService = {
  id: "db",
  name: "Mysql",
  type: "rds",
  engine: "mysql",
  engineVersion: "8",
  instanceSize: "db.t4g.micro",
  port: 3306,
  storage: 20,
  private: false
};
const databaseEnvVariables = {
  DATABASE_URL: {
    fromService: { id: "db", value: "dbConnectionString" }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  databaseEnvVariables,
  flightcontrolConfig,
  mysqlDatabaseService,
  postgresDatabaseService
});
