// @redwoodjs/vite/dist/cjs/rsc/rscStudioHandlers.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var rscStudioHandlers_exports = {};
__export(rscStudioHandlers_exports, {
  sendRscFlightToStudio: () => sendRscFlightToStudio
});
module.exports = __toCommonJS(rscStudioHandlers_exports);
var import_node_http = __toESM(require("node:http"), 1);
var import_project_config = require("@redwoodjs/project-config");
var import_rscRenderer = require("./rscRenderer.js");
const isTest = () => {
  return process.env.NODE_ENV === "test";
};
const isDevelopment = () => {
  return process.env.NODE_ENV !== "production" && !isTest();
};
const isStudioEnabled = () => {
  return (0, import_project_config.getRawConfig)()["studio"] !== void 0;
};
const shouldSendToStudio = () => {
  return isStudioEnabled() && !isDevelopment();
};
const getStudioPort = () => {
  return (0, import_project_config.getConfig)().studio.basePort;
};
const processRenderRscStream = async (readable) => {
  return new Promise((resolve, reject) => {
    const chunks = [];
    const writable = new WritableStream({
      write(chunk) {
        chunks.push(chunk);
      },
      close() {
        resolve(Buffer.concat(chunks).toString("utf8"));
      }
    });
    readable.pipeTo(writable).catch((error) => reject(error));
  });
};
const postFlightToStudio = (payload, metadata) => {
  if (shouldSendToStudio()) {
    const base64Payload = Buffer.from(payload).toString("base64");
    const encodedMetadata = Buffer.from(JSON.stringify(metadata)).toString(
      "base64"
    );
    const jsonBody = JSON.stringify({
      flight: {
        encodedPayload: base64Payload,
        encoding: "base64",
        encodedMetadata
      }
    });
    const options = {
      hostname: "localhost",
      port: getStudioPort(),
      path: "/.redwood/functions/rsc-flight",
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(jsonBody)
      }
    };
    const req = import_node_http.default.request(options, (res) => {
      res.setEncoding("utf8");
    });
    req.on("error", (e) => {
      console.error(
        `An error occurred sending the Flight Payload to Studio: ${e.message}`
      );
    });
    req.write(jsonBody);
    req.end();
  }
};
const createStudioFlightHandler = (readable, metadata) => {
  if (shouldSendToStudio()) {
    processRenderRscStream(readable).then((payload) => {
      console.debug("Sending RSC Rendered stream to Studio");
      postFlightToStudio(payload, metadata);
      console.debug("Sent RSC Rendered stream to Studio", payload, metadata);
    }).catch((error) => {
      console.error("An error occurred getting RSC Rendered steam:", error);
    });
  } else {
    console.debug("Studio is not enabled");
  }
};
const sendRscFlightToStudio = async (input) => {
  if (!shouldSendToStudio()) {
    console.debug("Studio is not enabled");
    return;
  }
  const { rscId, rsaId, args, basePath, req, handleError } = input;
  try {
    const startedAt = Date.now();
    const start = performance.now();
    const readable = await (0, import_rscRenderer.renderRscToStream)({ rscId, rsaId, args });
    const endedAt = Date.now();
    const end = performance.now();
    const duration = end - start;
    const metadata = {
      rsc: {
        rscId,
        rsaId,
        args
      },
      request: {
        basePath,
        originalUrl: req.originalUrl,
        url: req.url,
        headers: req.headers
      },
      performance: {
        startedAt,
        endedAt,
        duration
      }
    };
    createStudioFlightHandler(readable, metadata);
  } catch (e) {
    if (e instanceof Error) {
      console.error("An error occurred rendering RSC and sending to Studio:", e);
      handleError(e);
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  sendRscFlightToStudio
});
