// @redwoodjs/cli/dist/commands/destroy/directive/directive.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var directive_exports = {};
__export(directive_exports, {
  builder: () => builder,
  command: () => command,
  description: () => description,
  handler: () => handler,
  tasks: () => tasks
});
module.exports = __toCommonJS(directive_exports);
var import_directive = require("../../generate/directive/directive");
var import_helpers = require("../helpers");
const description = "Destroy a directive";
const { command, handler, builder, tasks } = (0, import_helpers.createYargsForComponentDestroy)({
  componentName: "directive",
  filesFn: (args) => (0, import_directive.files)({ ...args, type: "validator" })
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  builder,
  command,
  description,
  handler,
  tasks
});
