// @redwoodjs/web/dist/apollo/links.js
import { Observable } from "@apollo/client/core/index.js";
import { setContext } from "@apollo/client/link/context/index.js";
import { ApolloLink } from "@apollo/client/link/core/index.js";
import { HttpLink } from "@apollo/client/link/http/index.js";
import { print } from "graphql/language/printer.js";
function createHttpLink(uri, httpLinkConfig, cookieHeader) {
  const headers = {};
  if (cookieHeader) {
    headers.cookie = cookieHeader;
  }
  return new HttpLink({
    uri,
    credentials: "include",
    ...httpLinkConfig,
    headers
  });
}
function enhanceError(operation, error) {
  const { operationName, query, variables } = operation;
  error.__RedwoodEnhancedError = {
    operationName,
    operationKind: query?.kind.toString(),
    variables,
    query: query && print(query)
  };
  return error;
}
function createUpdateDataLink() {
  return new ApolloLink((operation, forward) => {
    return new Observable((observer) => {
      forward(operation).subscribe({
        next(result) {
          if (result.errors) {
            result.errors.forEach((error) => {
              enhanceError(operation, error);
            });
          }
          observer.next(result);
        },
        error(error) {
          observer.error(enhanceError(operation, error));
        },
        complete: observer.complete.bind(observer)
      });
    });
  });
}
function createAuthApolloLink(authProviderType, headersFromFetchProvider) {
  return new ApolloLink((operation, forward) => {
    const { token } = operation.getContext();
    const authHeaders = token ? {
      "auth-provider": authProviderType,
      authorization: `Bearer ${token}`
    } : {};
    if (!token) {
      delete headersFromFetchProvider?.["auth-provider"];
    }
    operation.setContext(() => ({
      headers: {
        ...operation.getContext().headers,
        ...headersFromFetchProvider,
        // Duped auth headers, because we may remove the `FetchConfigProvider` at a later date.
        ...authHeaders
      }
    }));
    return forward(operation);
  });
}
function createTokenLink(getToken) {
  return setContext(async () => {
    const token = await getToken();
    return { token };
  });
}
function createFinalLink({
  userConfiguredLink,
  defaultLinks
}) {
  if (userConfiguredLink) {
    if (typeof userConfiguredLink === "function") {
      return userConfiguredLink(defaultLinks);
    } else {
      return userConfiguredLink;
    }
  }
  return ApolloLink.from(defaultLinks.map((l) => l.link));
}
export {
  createAuthApolloLink,
  createFinalLink,
  createHttpLink,
  createTokenLink,
  createUpdateDataLink
};
