// @redwoodjs/cli/dist/commands/destroy/service/service.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var service_exports = {};
__export(service_exports, {
  command: () => command,
  description: () => description,
  filesWithTemplateVars: () => filesWithTemplateVars,
  handler: () => handler,
  tasks: () => tasks
});
module.exports = __toCommonJS(service_exports);
var import_lib = require("../../../lib");
var import_schemaHelpers = require("../../../lib/schemaHelpers");
var import_service = require("../../generate/service/service");
var import_helpers = require("../helpers");
const filesWithTemplateVars = (templateVars) => {
  return (args) => (0, import_service.files)({ ...args, ...templateVars });
};
const { command, description, handler, tasks } = (0, import_helpers.createYargsForComponentDestroy)({
  componentName: "service",
  preTasksFn: import_schemaHelpers.verifyModelName,
  filesFn: filesWithTemplateVars({ ...(0, import_lib.getDefaultArgs)(import_service.builder), crud: true })
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  command,
  description,
  filesWithTemplateVars,
  handler,
  tasks
});
